Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs

```diff
@@ -1257,6 +1262,20 @@ public Task<Result<byte[]>> GetBlobAsync(OperationContext context, ContentHash h
             return GlobalStore.GetBlobAsync(context, hash);
         }
 
+        private async void OnContentLocationDatabaseInvalidationAsync(object sender, ContentLocationDatabase.DatabaseInvalidatedEventArgs e)
+        {
+            // If multiple threads fail at the same time (i.e. corruption error), this cheaply deduplicates the
+            // restores, avoids redundant logging.
+            using (_databaseInvalidationGate.AcquireSemaphore(0))

```

Comment:
>AcquireSemaphore [](start = 45, length = 16)

Maybe Monitor.TryEnter instead of SemaphoreSlim here. Do the other threads need to block or is it sufficient to just prevent reentrancy?

Reply:
But do we actually need it at all? ProcessStateAsync prevents re-entrancy.

---
In reply to: [327874599](https://github.com/microsoft/BuildXL/pull/923#discussion_r327874599) [](ancestors = 327874599)

Reply:
Looking at RunOutOfBandAsync it doesn't seem to be threadsafe.

---
In reply to: [327874969](https://github.com/microsoft/BuildXL/pull/923#discussion_r327874969) [](ancestors = 327874969,327874599)

Reply:
This prevents re-entrancy by forcing the other threads to do a non-blocking return. I forgot to add the 0 in the Wait inside, but that is the idea at least

---
In reply to: [327876278](https://github.com/microsoft/BuildXL/pull/923#discussion_r327876278) [](ancestors = 327876278,327874969,327874599)